File: tcp-ip/ch06/conn_client.cc
//
// Created by Tiancheng on 2024/8/19.
//

#include <arpa/inet.h>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <sys/socket.h>
#include <unistd.h>

#define BUF_SIZE 32

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Usage: %s <serverAddr> <serverPort>\n", argv[0]);
        exit(1);
    }

    //* 调用 socket 函数，创建 UDP socket 套接字
    int clientSocketFd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    if (clientSocketFd == -1) {
        printf("Error created socket\n");
    }

    sockaddr_in serverAddr{};
    serverAddr.sin_family = AF_INET;
    serverAddr.sin_addr.s_addr = inet_addr(argv[1]); // serverAddr
    serverAddr.sin_port = htons(atoi(argv[2]));      // serverPort

    //* 创建已连接 UDP 套接字
    //* 客户端调用 connect 函数，向服务器发送连接请求
    connect(clientSocketFd, (sockaddr *)&serverAddr, sizeof(serverAddr));

    char send[BUF_SIZE];
    char recv[BUF_SIZE];
    while (true) {
        fputs("Input: ", stdout);
        fgets(send, BUF_SIZE, stdin);
        // send[strlen(send) - 1] = '\0'; // 将 \n 替换为 \0
        // 客户端向服务器发送 UDP 数据报，客户端自动分配本机 IP 地址和端口
        write(clientSocketFd, send, strlen(send));
        // memset(recv, '\0', BUF_SIZE); //! 将 recv 中所有元素置为 '\0'
        //! 从 clientSocketFd 中读出 BUF_SIZE 个字节, 写入 recv 缓冲区
        ssize_t readBytes = read(clientSocketFd, recv, BUF_SIZE);
        if (readBytes <= 0) {
            break;
        }
        recv[readBytes] = '\0'; //! 隔离脏数据
        printf("Echo from server %s\n", recv);
    }
    close(clientSocketFd);
    return 0;
}

File: tcp-ip/ch06/go/conn_client/conn_client.go
package main

import (
	"fmt"
	"net"
	"os"
)

const BUF_SIZE = 30

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <serverAddr>:<serverPort>\n", os.Args[0])
		os.Exit(1)
	}
	serverAddrStr /* remoteAddr */ := os.Args[1]
	//* 客户端调用 net.ListenPacket 函数，监听远端（服务器）的 UDP 连接请求
	// listener, err := net.ListenPacket("udp", port)
	serverAddr /* remoteAddr */, _ := net.ResolveUDPAddr("udp", serverAddrStr)
	//! 建立 UDP 会话，推荐
	conn, err := net.DialUDP("udp", nil /* 自动分配端口 */, serverAddr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "[ERROR] %s\n", err.Error())
		os.Exit(1)
	}
	send := make([]byte, BUF_SIZE)
	recv := make([]byte, BUF_SIZE)
	for {
		fmt.Print("Input: ")
		/* nBytes, err := */ fmt.Scanln(&send)
		//! 已连接的 UDP 套接字 - Write, Read
		conn.Write(send) // 已连接的 UDP 套接字发送 UDP 数据报时无需指定远端 IP 地址
		readBytes, err := conn.Read(recv)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[ERROR] %s\n", err.Error())
		}
		// 等价于 fmt.Printf("Echo from server %s\n", recv[:readBytes])
		fmt.Printf("Echo from server %s\n", recv[:readBytes])
	}
}
